docs/polymorphic-table-types.md

```diff
@@ -0,0 +1,198 @@
+
+### General Indexer Pattern
+
+At an application level, the current workaround for this is to wrap the table in a function getter like so:
+
+```lua
+-- private
+local foo = {} :: { [F<any>]: G<any> }
+
+-- public
+local function getFoo<T>(k: F<T>): G<T>
+	return foo[k]
+end
+
+local function setFoo<T>(k: F<T>, v: G<T>)
+	foo[k] = v
+end
+
+local iterFoo = pairs(foo)
+```
+
+This has five immediate consequences.
+
+1. This pattern extends to every table of this nature, increasing surface area, duplication, and maintenance.
+2. In the event the functions don't get inlined, we incur an extra performance penalty per read and write, disallowing certain applications type-safety.
+
+If we are using a function then one might try and take advantage of the extra indirection by hiding the table implementation to prevent misuse. This raises the following consequences:
+
+3. The implementation is opaque and can be modified later without worry.
+4. Mutations require yet another function, a setter, with the cons listed before.
+5. Iterations requires yet another function, or at least stored iterator, reinforcing the cons listed before.
+
+In short, tables are mutable and iterable, functions are not.
+
+At this point if one is not satisfied with the extra baggage that goes behind a type-safe access or write of this nature, they will simply lie to the type checker or compromise and move on. This is exactly what we wish to avoid.
+
+This proposal targets users already neck-deep in the type-system supporting libraries and tools, unable to achieve safety without compromising performance or their api and affecting thousands of users, or the occasional type-savvy user getting that perfect type for their data structure and feeling accomplished.
+
+### Specific Indexer Pattern
+
+We'll demonstrate getting components from an entity using a minified ECS api from a library currently in production. This library depends on accessing components from tables using their constructors as indexers. The constructors contain the type information of the component, so it is theoretically possible to infer the type of the resulting component from the constructor, but this is not possible as of now because polymorphic tables don't exist.
+
+```lua
+-- <C>(() -> C) -> Factory<C>
+world.factory
+
+-- <'Foo'>(() -> 'Foo') -> Factory<'Foo'> ==> Factory<'Foo'>
+local A = world.factory(function()
+	return 'Foo'
+end)
+
+-- <'Bar'>(() -> 'Bar') -> Factory<'Bar'> ==> Factory<'Bar'>
+local B = world.factory(function()
+	return 'Bar'
+end)
+
+-- number
+local e = world.entity()
+
+-- 'Foo'
+A.add(e)
+
+-- 'Bar'
+B.add(e)
+
+-- { [Factory<any>]: any? } 	(no possible way to extract the generic)
+local components = world.get(e)
```

Comment:
Why can't `world.get(e)` return a function like `<T>(Factory<T>) -> T?`? Have you actually measured that this has a serious negative performance impact? It seems generally undesirable to me to have an ECS provide a mutable table of components for an entity as a result of a lookup, so it feels like there'd need to be a good reason to prefer this over an API that encapsulates the internal representation of components.